### PR TITLE
planner,variable: disable tidb_enable_reuse_chunk when reading json/text column

### LINKED ISSUE #41366: tidb_enable_reuse_chunk increases the risk of OOM when reading json/text column
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
When a sql reads json/text column from a table, the memory consumption is too large if enabling `tidb_enable_reuse_chunk`.
NPjt53wIlW

The json columns will be referred until the SQL is finished when enabling `tidb_enable_reuse_chunk`.

I think we do not need to reuse the columns of type json/text, they consume too large memory.
Considering the implementation, if we only reject the reuse of json/text column, the memory still cannot be GCed because `tidb_enable_chunk_rpc` is on by default. Columns with different type are stored in the same slice when reading from rpc.

We can disable the tidb_enable_chunk_rpc for given SQL if it needs to read json/text column, and enable tidb_enable_chunk_rpc after this SQL is finished.

### 2. What did you expect to see? (Required) 
Memory usage is stable.

### 3. What did you see instead (Required)
There is a risk of OOM.

### 4. What is your TiDB version? (Required)

master

### PR DESCRIPTION
**Pull request by bob34007**
### What problem does this PR solve?

Issue Number: close #41366

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/optimizer.go [hunk 1]
```diff
@@ -1220,3 +1221,70 @@ func init() {
...
+			mysql.TypeBlob, mysql.TypeJSON:
+			return true
+		case mysql.TypeVarchar:
+			if column.RetType.GetFlen() > 1000 {
+				return true
+			}
+		}
+	}
+	return false
+}
```
Comment (XuHuaiyu) [inline planner/core/optimizer.go:1290]:
```suggestion
// checkOverlongColType Check if read field type is too long.
func checkOverlongColType(sctx sessionctx.Context, plan PhysicalPlan) bool {
	switch plan.(type) {
	case *PhysicalTableReader, *PhysicalIndexReader, *PhysicalIndexLookUpReader, *PhysicalIndexMergeReader, *PointGet:
		if existsOverlongType(plan.Schema().Columns) {
			sctx.GetSessionVars().ClearAlloc(nil, false)
			return true
		}
	}
	return false
}

func existsOverlongType(columns []*expression.Column) bool {
	for _, column := range columns {
		switch column.RetType.GetType() {
		case mysql.TypeTinyBlob, mysql.TypeMediumBlob, mysql.TypeLongBlob,
			mysql.TypeBlob, mysql.TypeJSON:
			return true
		case mysql.TypeVarString, mysql.TypeVarchar:
			if column.RetType.GetFlen() > 1000 {
				return true
			}
		}
	}
	return false
}
```

File: sessionctx/variable/session.go [hunk 2]
```diff
@@ -1427,6 +1427,16 @@ func (s *SessionVars) SetAlloc(alloc chunk.Allocator) {
 	s.ChunkPool.Alloc = alloc
 }
 
+// CheckAlloc check if alloc is enable
+func (s *SessionVars) CheckAlloc() bool {
```
Comment (XuHuaiyu) [inline sessionctx/variable/session.go:1431]:
```suggestion
func (s *SessionVars) IsAllocValid() bool {
```

File: planner/core/optimizer.go [hunk 3]
```diff
@@ -1220,3 +1221,70 @@ func init() {
 	DefaultDisabledLogicalRulesList = new(atomic.Value)
 	DefaultDisabledLogicalRulesList.Store(set.NewStringSet())
 }
+
+// checkCanUseReuseChunk Check if chunk reuse can be used.
+func checkCanUseReuseChunk(ctx context.Context, sctx sessionctx.Context, plan PhysicalPlan) {
```
Comment (XuHuaiyu) [inline planner/core/optimizer.go:1226]:
```suggestion
func disableReuseChunkIfNeeded(sctx sessionctx.Context, plan PhysicalPlan) {
```

File: planner/core/optimizer.go [hunk 4]
```diff
@@ -399,6 +399,7 @@ func postOptimize(ctx context.Context, sctx sessionctx.Context, plan PhysicalPla
 	handleFineGrainedShuffle(ctx, sctx, plan)
 	propagateProbeParents(plan, nil)
 	countStarRewrite(plan)
+	checkCanUseReuseChunk(ctx, sctx, plan)
```
Comment (XuHuaiyu) [inline planner/core/optimizer.go:402]:
```suggestion
	disableReuseChunkIfNeeded(sctx, plan)
```

File: planner/core/optimizer.go [hunk 5]
```diff
@@ -1220,3 +1221,57 @@ func init() {
 	DefaultDisabledLogicalRulesList = new(atomic.Value)
 	DefaultDisabledLogicalRulesList.Store(set.NewStringSet())
 }
+
+// checkCanUseReuseChunk Check if chunk reuse can be used.
```
Comment (XuHuaiyu) [inline planner/core/optimizer.go:1225]:
```suggestion
```

**Comment by XuHuaiyu** [comment]
@keeplearning20221 Build failed, please fix the ci.

**Comment by keeplearning20221** [comment]
/retest

> @keeplearning20221 Build failed, please fix the ci.
@bob34007

**Review by XuHuaiyu — APPROVED** [review]

**Review by wshwsh12 — APPROVED** [review]

**Review by XuHuaiyu — APPROVED** [review]

**Pull request merged**